core/vm/interpreter.go

```diff
@@ -204,17 +207,6 @@ func (in *EVMInterpreter) Run(contract *Contract, input []byte, readOnly bool) (
 		} else if sLen > operation.maxStack {
 			return nil, &ErrStackOverflow{stackLen: sLen, limit: operation.maxStack}
 		}
-		// If the operation is valid, enforce write restrictions
```

Comment:
So this is changing the order of operations a bit. Like one concrete change is the tracer's `CaptureState` will be called even on failure. Also memory will be resized, but I guess that's moot since the error will cause the call frame to fail. Another possibility is EVM returning a different error message if there are multiple failures (i.e. readonly violation AND not enough gas for memory expansion)

Reply:
Absolutely, this is one of my concerns, see the discussion on #23968. The order of evaluation for the most part is not covered by state tests, but in the long term I would hope it will be covered.

Reply:
Ah right missed that convo. Maybe time to revive the "trace after op execution" discussion

Reply:
> The order of evaluation for the most part is not covered by state tests

It's not so much that it isn't covered -- the fact is that the order is opaque to state tests, in nearly all situations.

Reply:
We had ideated to expose error codes from execution and include those in the state tests. That would be a way to ensure some common order.